kornia/enhance/jpeg.py

```diff
@@ -0,0 +1,566 @@
+    # Perform chroma upsampling
+    image_cb = _chroma_upsampling(image_cb)
+    image_cr = _chroma_upsampling(image_cr)
+    # Back to [0, 1] pixel-range
+    image_ycbcr: Tensor = torch.stack((image_y, image_cb, image_cr), dim=1) / 255.0
+    # Convert back to RGB space.
+    rgb_decoded: Tensor = ycbcr_to_rgb(image_ycbcr)
+    return rgb_decoded
+
+
+def diff_jpeg(
+    image_rgb: Tensor,
+    jpeg_quality: Tensor,
+    quantization_table_y: Tensor = QUANTIZATION_TABLE_Y,
+    quantization_table_c: Tensor = QUANTIZATION_TABLE_C,
+) -> Tensor:
+    r"""Differentiable JPEG encoding-decoding module.
+
+    Based on [1, 2], we perform differentiable JPEG encoding-decoding as follows:
+
+    .. math::
+
+        \text{JPEG}_{\text{diff}}(I, q, QT_{y}, QT_{c}) = \hat{I}
+
+    Where:
+       - :math:`I` is the original image to be coded.
+       - :math:`q` is the JPEG quality controling the compression strength.
+       - :math:`QT_{y}` is the luma quantization table.
+       - :math:`QT_{c}` is the chroma quantization table.
+       - :math:`\hat{I}` is the resulting JPEG encoded-decoded image.
+
+    Notes:
+        The input (and output) pixel range is :math:`[0, 1]`. In case you want to handle normalized images you are
+        required to first perform denormalization followed by normalizing the output images again.
+        Note, this implementation models the encoding-decoding mapping of JPEG in a differentiable setting,
+        howerver, does not allow to excess the JPEG-coded byte file itself. For more details please refer to [1].
+
+    Reference:
+        [1] https://arxiv.org/abs/2309.06978
+        [2] https://machine-learning-and-security.github.io/papers/mlsec17_paper_54.pdf
+
+    Args:
+        image_rgb: the RGB image to be coded.
+        jpeg_quality: JPEG quality in the range :math:`[0, 99]` controling the compression strength.
+        quantization_table_y: quantization table for Y channel. Default: standard quantization table.
+        quantization_table_c: quantization table for C channels. Default: standard quantization table.
+
+    Shape:
+        - image_rgb: :math:`(B, 3, H, W)`.
+        - jpeg_quality: :math:`(1)` or :math:`(B)`.
+        - quantization_table_y: :math:`(8, 8)` or :math:`(B, 8, 8)`.
+        - quantization_table_c: :math:`(8, 8)` or :math:`(B, 8, 8)`.
+
+    Return:
+        JPEG coded image of the shape :math:`(B, 3, H, W)`
+
+    Example:
+        >>> img = torch.rand(3, 3, 64, 64, requires_grad=True, dtype=torch.float)
+        >>> jpeg_quality = torch.tensor((99.0, 25.0, 1.0), requires_grad=True)
+        >>> img_jpeg = diff_jpeg(img, jpeg_quality)
+        >>> img_jpeg.sum().backward()
+    """
+    # Check that inputs are tensors
+    KORNIA_CHECK_IS_TENSOR(image_rgb)
+    KORNIA_CHECK_IS_TENSOR(jpeg_quality)
+    KORNIA_CHECK_IS_TENSOR(quantization_table_y)
+    KORNIA_CHECK_IS_TENSOR(quantization_table_c)
+    # Check shape of inputs
+    KORNIA_CHECK_SHAPE(image_rgb, ["B", "3", "H", "W"])
```

Comment:
would be great (or in a separated PR) to support out of the box non-batched images 3xHxW

Reply:
Let's do it in a separate PR if it's fine.